sycl/source/detail/graph_impl.hpp

```diff
@@ -235,50 +178,54 @@ class node_impl {
     return nullptr;
   }
 
-  /// Tests is the caller is similar to Node
+  /// Tests if the caller is similar to Node, this is only used for testing.
+  /// @param Node The node to check for similarity.
+  /// @param CompareContentOnly Skip comparisons related to graph structure,
+  /// compare only the type and command groups of the nodes
   /// @return True if the two nodes are similar
-  bool isSimilar(std::shared_ptr<node_impl> Node) {
-    if (MSuccessors.size() != Node->MSuccessors.size())
-      return false;
+  bool isSimilar(std::shared_ptr<node_impl> Node,
+                 bool CompareContentOnly = false) {
+    if (!CompareContentOnly) {
+      if (MSuccessors.size() != Node->MSuccessors.size())
+        return false;
 
-    if (MPredecessors.size() != Node->MPredecessors.size())
+      if (MPredecessors.size() != Node->MPredecessors.size())
+        return false;
+    }
+    if (MCGType != Node->MCGType)
       return false;
 
-    if (*this == *Node.get())
-      return true;
-
-    return false;
-  }
-
-  /// Recursive traversal of successor nodes checking for
-  /// equivalent node successions in Node
-  /// @param Node pointer to the starting node for structure comparison
-  /// @return true is same structure found, false otherwise
-  bool checkNodeRecursive(std::shared_ptr<node_impl> Node) {
-    size_t FoundCnt = 0;
-    for (std::shared_ptr<node_impl> SuccA : MSuccessors) {
-      for (std::shared_ptr<node_impl> SuccB : Node->MSuccessors) {
-        if (isSimilar(Node) && SuccA->checkNodeRecursive(SuccB)) {
-          FoundCnt++;
-          break;
-        }
-      }
+    switch (MCGType) {
```

Comment:
I believe those are handled in their respective PRs, so I would prefer to keep them there (with the caveat that yes they will have to make changes based on this refactoring). Especially since those are blocked by UR and might not be in for a while.